apps/language_server/test/experimental/provider/handlers/goto_definition_test.exs

```diff
@@ -0,0 +1,214 @@
+defmodule ElixirLS.Experimental.Provider.Handlers.GotoDefinitionTest do
+  use ExUnit.Case, async: true
+
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Requests.GotoDefinition
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Responses
+  alias ElixirLS.LanguageServer.Experimental.Provider.Env
+  alias ElixirLS.LanguageServer.Experimental.Provider.Handlers
+  alias ElixirLS.LanguageServer.Experimental.SourceFile
+  alias ElixirLS.LanguageServer.Experimental.SourceFile.Conversions
+
+  alias ElixirLS.LanguageServer.Fixtures.LspProtocol
+  alias ElixirLS.LanguageServer.Test.FixtureHelpers
+
+  import LspProtocol
+  import ElixirLS.Test.TextLoc, only: [annotate_assert: 4]
+
+  setup do
+    {:ok, _} = start_supervised(SourceFile.Store)
+    :ok
+  end
+
+  def request(file_path, line, char) do
+    uri = Conversions.ensure_uri(file_path)
+
+    params = [
+      text_document: [uri: uri],
+      position: [line: line, character: char]
+    ]
+
+    with {:ok, contents} <- File.read(file_path),
+         :ok <- SourceFile.Store.open(uri, contents, 1),
+         {:ok, req} <- build(GotoDefinition, params) do
+      GotoDefinition.to_elixir(req)
+    end
+  end
+
+  def handle(request) do
+    Handlers.GotoDefinition.handle(request, Env.new())
+  end
+
+  defp arrange_referenced_file do
+    "references_referenced.ex"
+    |> FixtureHelpers.get_path()
+    |> Conversions.ensure_uri()
+  end
+
+  test "find definition remote function call" do
+    referenced_uri = arrange_referenced_file()
```

Comment:
ok, now I see what you're doing. Why not make a setup block?
```elixir
def with_rerenced_file(_) do
 path = FixtureHelpers.get_path("references_referenced.ex")
 uri = Conversions.ensure_uri(path)
  {:ok, file_uri: uri, file_path: path}
  end
 ```
Then: 
```elixir 
describe "when a file contains references" do 
  setup [:with_referenced_file] 
  
  test "something that needs the uri", %{file_uri: uri} do 
    ...
   end  

  test "something that needs the path", %{file_path: path} do 
    ...
   end  
end
```

Reply:
First of all I am not a test fixture fan. I understand that using setup is possible to reduce duplicate code, but this place is not necessary.

The reason is as follows.

The code logic now looks like this.

1. Arrange the referenced file and the reference file
2. annotate something
3. Act: handle(request)
4. Assertions

No deps for any tests, all the information in that test.

But if we use a setup, We will let some important information too far from the test(we have to jump to the with_referenced_file if we want to figure out how these two files interact).

the logic become:

%{file_uri: uri}

1. Arrange a file
2. annotate something
3. Act: handle(request)
4. Assertions

It's easy to get confused by look at one test.

FYI: Chapter 3.3.3 of this [book](https://www.manning.com/books/unit-testing) has a similar point, and I agree with him.

Reply:
I just don't like overly repetitive setup code, and that's what setup blocks are for, they're not the same thing as fixtures, which you're using regardless. I like tests too, and you should really focus on letting the code inside of a test do the testing work; it makes things clearer, especially when your setup blocks are fine-grained (as was the one I showed above).

I also wasn't a fan of the `arrange_` prefix; though I'd drop the `get_` prefix on the new name, since you're really not getting something , it's just the uri of the file that's being built each time. Furthermore, both the things emitted from the setup block (`path` and `uri`) are really the same thing, just different formats, and now you're calling the fixture twice rather than once.  That's a lot of cognitive load to not use a setup block, which would still reduce the LOC by two per test right now. 

Please use the setup block here.

Reply:
Thank you for your detailed response, I think I have fully understood your arguments and contentions. Reducing the cognitive load of the code is also my goal.